Add ViT finetuning post (#227)
huggingface/blog: fine-tune-vit.md

            <span class="fullname">Nate Raw</span>
        </div>
    </a>
</div>

<script async defer src="https://unpkg.com/medium-zoom-element@0/dist/medium-zoom-element.min.js"></script>

<a target="_blank" href="https://colab.research.google.com/github/nateraw/huggingface-hub-examples/blob/main/vit_image_classification_explained.ipynb">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

Just as transformers-based models have revolutionized NLP, we're now seeing an explosion of papers applying them to all sorts of other domains. One of the most revolutionary of these was the Vision Transformer (ViT), which was introduced in [June 2021](https://arxiv.org/abs/2010.11929) by a team of researchers at Google Brain.

This paper explored how you can tokenize images, just as you would tokenize sentences, so that they can be passed to transformer models for training. Its quite a simple concept, really...

1. Split an image into a grid of sub-image patches
1. Embed each patch with a linear projection
1. Each embedded patch becomes a token, and the resulting sequence of embedded patches is the sequence you pass to the model.

> 

Author: Fun fact, in Markdown you can use a sequence of numbers that don't increase, and markdown will render them correctly for you. Here's what it looks like in moon landing:

---

<img width="925" alt="Screen Shot 2022-02-10 at 11 33 11 AM" src="https://user-images.githubusercontent.com/32437151/153452694-c2ffb95b-90bf-4b83-a7a6-63d81f12c26f.png">

> ![vit_figure.png](https://raw.githubusercontent.com/google-research/vision_transformer/main/vit_figure.png)

Author: Yes - and a note for myself...this should be a local image not the URL. I downloaded it, so it should be opening directly instead.

It turns out that once you've done the above, you can pre-train and finetune transformers just as you're used to with NLP tasks. Pretty sweet 😎.

---

In this notebook, we'll walk through how to leverage 🤗 `datasets` to download and process image classification datasets, and then use them to fine-tune a pre-trained ViT with 🤗 `transformers`. 

To get started, lets first install both those packages.

```bash
pip install datasets transformers
```

## Load a dataset

Let's start by loading a small image classification dataset and taking a look at its structure.

We'll use the [`beans`](https://huggingface.co/datasets/beans) dataset, which is a collection of pictures of healthy and unhealthy bean leaves. 🍃

```python
from datasets import load_dataset

ds = load_dataset('beans')
ds
```

> Let's take a look at the 400th example from the `'train'` split from the beans dataset. You'll notice each example from the dataset has 3 features:

Author: Disagree on this one - more of a personal voice thing here.

1. `image`: A PIL Image
1. `image_file_path`: The `str` path to the image file that was loaded as `image`

> 1. `labels`: A [`datasets.ClassLabel`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=classlabel#datasets.ClassLabel) feature, which we'll see as an integer representation of the label for a given example. (Later we'll see how to get the string class names, don't worry)

Author: hmm, I see what you're trying to change here. I'll edit this one myself. Will mark as resolved when I do so.

</figure>

From what I'm seeing, 
- Angular Leaf Spot: Has irregular brown patches
- Bean Rust:  Has circular brown spots surrounded with a white-ish yellow ring
- Healthy: ...looks healthy. 🤷‍♂️

## Loading ViT Feature Extractor

Now that we know what our images look like and have a better understanding of the problem we're trying to solve, let's see how we can prepare these images for our model. 

When ViT models are trained, specific transformations are applied to images being fed into them. Use the wrong transformations on your image and the model won't be able to understand what it's seeing! 🖼 ➡️ 🔢

To make sure we apply the correct transformations, we will use a [`ViTFeatureExtractor`](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=classlabel#datasets.ClassLabel.int2str) initialized with a configuration that was saved along with the pretrained model we plan to use. In our case, we'll be using the [google/vit-base-patch16-224-in21k](https://huggingface.co/google/vit-base-patch16-224-in21k) model, so lets load its feature extractor from the 🤗 Hub.

```python
from transformers import ViTFeatureExtractor

> model_name_or_path = 'google/vit-base-patch16-224-in21k'

Author: I think I'll leave this one as-is. Somewhat of a personal preference 🤷

        0.5,
        0.5
      ],
      "resample": 2,
      "size": 224
    }

To process an image, simply pass it to the feature extractor's call function. This will return a dict containing `pixel values`, which is the numeric representation of your image that we'll pass to the model.

We get a numpy array by default, but if we add the `return_tensors='pt'` argument, we'll get back `torch` tensors instead.

```python
feature_extractor(image, return_tensors='pt')
```

Should give you something like...

    {
      'pixel_values': tensor([[[[ 0.2706,  0.3255,  0.3804,  ...]]]])
    }

...where the shape of the tensor is `(1, 3, 224, 224)`.

## Processing the Dataset

Now that we know how to read in images and transform them into inputs, let's write a function that will put those two things together to process a single example from the dataset.

```python
def process_example(example):
    inputs = feature_extractor(example['image'], return_tensors='pt')
    inputs['labels'] = example['labels']
    return inputs
```

```python
process_example(ds['train'][0])
```

    {
      'pixel_values': tensor([[[[-0.6157, -0.6000, -0.6078,  ..., ]]]]),
      'labels': 0
    }

> While we could call `ds.map` and apply this to every example at once, this can be very slow, especially if you use a larger dataset. Instead, we'll apply a ***transform*** to the dataset. Transforms are only applied to examples as you index them.

Reviewer: ```suggestion
While you could call `ds.map` and apply this to every example at once, this can be very slow, especially if you use a larger dataset. Instead, you can apply a ***transform*** to the dataset. Transforms are only applied to examples as you index them.
```

> First, though, we'll need to update our last function to accept a batch of data, as that's what `ds.with_transform` expects.

Reviewer: ```suggestion
First, though, you'll need to update the last function to accept a batch of data, as that's what `ds.with_transform` expects.
```

```python
ds = load_dataset('beans')

def transform(example_batch):
    # Take a list of PIL images and turn them to pixel values
    inputs = feature_extractor([x for x in example_batch['image']], return_tensors='pt')

    # Don't forget to include the labels!
    inputs['labels'] = example_batch['labels']
    return inputs
```

> We can directly apply this to our dataset using `ds.with_transform(transform)`.

Reviewer: ```suggestion
You can directly apply this to the dataset using `ds.with_transform(transform)`.
```

```python
prepared_ds = ds.with_transform(transform)
```

> Now, whenever we get an example from the dataset, our transform will be 

Reviewer: ```suggestion
Now, whenever you get an example from the dataset, the transform will be 
```

applied in real time (on both samples and slices, as shown below)

```python
prepared_ds['train'][0:2]
```

This time, the resulting `pixel_values` tensor will have shape `(2, 3, 224, 224)`.

    {
      'pixel_values': tensor([[[[-0.6157, -0.6000, -0.6078,  ..., ]]]]),
      'labels': [0, 0]
    }

# Training and Evaluation

> The data is processed and we are ready to start setting up the training pipeline. We will make use of 🤗's Trainer, but that'll require us to do a few things first:

Reviewer: ```suggestion
The data is processed and you are ready to start setting up the training pipeline. This blog post uses 🤗's Trainer, but that'll require to do a few things first:
```

- Define a collate function.

> - Define an evaluation metric. During training, the model should be evaluated on its prediction accuracy. We should define a compute_metrics function accordingly.

Reviewer: ```suggestion
- Define an evaluation metric. During training, the model should be evaluated on its prediction accuracy. You should define a `compute_metrics` function accordingly.
```

- Load a pretrained checkpoint. We need to load a pretrained checkpoint and configure it correctly for training.

- Define the training configuration.

> After having fine-tuned the model, we will correctly evaluate it on the evaluation data and verify that it has indeed learned to correctly classify our images.

Reviewer: ```suggestion
After fine-tuning the model, you will correctly evaluate it on the evaluation data and verify that it has indeed learned to correctly classify our images.
```

### Define our data collator

Batches are coming in as lists of dicts, so we just unpack + stack those into batch tensors.

We return a batch `dict` from our `collate_fn` so we can simply `**unpack` the inputs to our model later. ✨

```python
import torch

def collate_fn(batch):
    return {
        'pixel_values': torch.stack([x['pixel_values'] for x in batch]),
        'labels': torch.tensor([x['labels'] for x in batch])
    }
```

### Define an evaluation metric

> Here, we load the [accuracy](https://huggingface.co/metrics/accuracy) metric from `datasets`, and then write a function that takes in a model prediction + computes the accuracy.

Reviewer: ```suggestion
The [accuracy](https://huggingface.co/metrics/accuracy) metric from `datasets` can easily be used to compare the predictions with the labels.
```

```python
import numpy as np
from datasets import load_metric

metric = load_metric("accuracy")
def compute_metrics(p):
    return metric.compute(predictions=np.argmax(p.predictions, axis=1), references=p.label_ids)
```

> Now we can load our pretrained model. We'll add `num_labels` on init to make sure the model creates a classification head with the right number of units. We'll also include the `id2label` and `label2id` mappings so we have human readable labels in the 🤗 hub widget if we choose to `push_to_hub`.

Reviewer: ```suggestion
Let's load the pretrained model. We'll add `num_labels` on init so the model creates a classification head with the right number of units. We'll also include the `id2label` and `label2id` mappings to have human-readable labels in the Hub widget.
```

```python
from transformers import ViTForImageClassification

labels = ds['train'].features['labels'].names

model = ViTForImageClassification.from_pretrained(
    model_name_or_path,
    num_labels=len(labels),
    id2label={str(i): c for i, c in enumerate(labels)},
    label2id={c: str(i) for i, c in enumerate(labels)}
)
```

> We're almost ready to train! The last thing we'll do before that is set up the training configuration by defining [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.16.2/en/main_classes/trainer#transformers.TrainingArguments).

Reviewer: ```suggestion
Almost ready to train! The last thing needed before that is to set up the training configuration by defining [`TrainingArguments`](https://huggingface.co/docs/transformers/v4.16.2/en/main_classes/trainer#transformers.TrainingArguments).
```

Most of these are pretty self-explanatory, but one that is quite important here is `remove_unused_columns=False`. This one will drop any features not used by the model's call function. By default it's `True` because usually its ideal to drop unused feature columns, as it makes it easier to unpack inputs into the model's call function. But, in our case, we need the unused features ('image' in particular) in order to create 'pixel_values'.

> What I'm trying to say is that you'll have a bad time if you forget to set `remove_unused_columns=False`.

Reviewer: 🤣

  push_to_hub=False,
  report_to='tensorboard',
  load_best_model_at_end=True,
)
```

Now, all instances can be passed to Trainer and we are ready to start training!

```python
from transformers import Trainer

trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=collate_fn,
    compute_metrics=compute_metrics,
    train_dataset=prepared_ds["train"],
    eval_dataset=prepared_ds["validation"],
    tokenizer=feature_extractor,
)
```

### Train 🚀

```python
train_results = trainer.train()
trainer.save_model()
trainer.log_metrics("train", train_results.metrics)
trainer.save_metrics("train", train_results.metrics)
trainer.save_state()
```

### Evaluate 📊

```python
metrics = trainer.evaluate(prepared_ds['validation'])
trainer.log_metrics("eval", metrics)
trainer.save_metrics("eval", metrics)
```

Here were my evaluation results - Cool beans! Sorry, had to say it.

    ***** eval metrics *****
      epoch                   =        4.0
      eval_accuracy           =      0.985
      eval_loss               =     0.0637
      eval_runtime            = 0:00:02.13
      eval_samples_per_second =     62.356
      eval_steps_per_second   =       7.97

> Finally, if you want, you can push your model up to the hub. Here, we'll push it up if you specified `push_to_hub=True` in the training configuration.

Reviewer: ```suggestion
Finally, if you want, you can push your model to the Hub. Here, we'll push it up if you specified `push_to_hub=True` in the training configuration.
```

```python
kwargs = {
    "finetuned_from": model.config._name_or_path,
    "tasks": "image-classification",
    "dataset": 'beans',
    "tags": ['image-classification'],
}

if training_args.push_to_hub:
    trainer.push_to_hub('🍻 cheers', **kwargs)
else:
    trainer.create_model_card(**kwargs)
```

> The resulting model has been shared to [nateraw/vit-base-beans](https://huggingface.co/nateraw/vit-base-beans). I'm assuming you don't have pictures of bean leaves laying around, but if you do, give it a try! 🚀

Reviewer: You could add some example images with https://huggingface.co/docs/hub/model-repos#how-can-i-control-my-models-widget-example-input
